### PR TITLE
Add stochastic/noisy act_on replacement for state vectors

### PR DESCRIPTION
**Pull request by asciineuron**
@rmlarose this could be interesting, let me know what you think.

Our problem with noisy simulation is cirq doesn't handle it right for state vectors. Their `cirq.Simulator` can do stochastic sampling but this is at the simulator level somewhere, rather than `act_on`.

I implemented here this stochastic technique with act on. It basically decomposes any noise channel into its kraus operators. We can turn these into (non-unitary) gates, apply each to a copy of our state vector state, and compute the corresponding probabilities. We then sample one of these as our return value and renormalize the state vector.

Basically the same procedure under the hood as measurement operators but with the kraus decomposition.

quantum-measurement

### CODE REVIEW

**Comment by rmlarose** [comment]
`cirq.protocols.act_on` already does this, no?

```python
import cirq

qubits = cirq.LineQubit.range(2)
state = cirq.StateVectorSimulationState(qubits=qubits, initial_state=0)

op = cirq.H.on(qubits[0])
assert cirq.has_unitary(op)
cirq.protocols.act_on(op, state)
print(cirq.dirac_notation(state.target_tensor.reshape(4)))

op = cirq.BitFlipChannel(p=0.5).on(qubits[0])
assert not cirq.has_unitary(op) and cirq.has_mixture(op)
cirq.protocols.act_on(op, state)
print(cirq.dirac_notation(state.target_tensor.reshape(4)))

op = cirq.AmplitudeDampingChannel(0.5).on(qubits[0])
assert not cirq.has_unitary(op) and not cirq.has_mixture(op) and cirq.has_kraus(op)
cirq.protocols.act_on(op, state)
print(cirq.dirac_notation(state.target_tensor.reshape(4)))
```

**Comment by rmlarose** [comment]
Great, let's keep it open to track it.
